plugins/inputs/mavlink/mavlink.go

```diff
@@ -0,0 +1,123 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package mavlink
+
+import (
+	_ "embed"
+	"fmt"
+
+	"github.com/chrisdalke/gomavlib/v3"
+	"github.com/chrisdalke/gomavlib/v3/pkg/dialects/ardupilotmega"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/internal/choice"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+// Plugin state
+type Mavlink struct {
+	// Config param
+	FcuURL                 string   `toml:"fcu_url"`
+	SystemID               uint8    `toml:"system_id"`
+	MessageFilter          []string `toml:"message_filter"`
+	StreamRequestEnable    bool     `toml:"stream_request_enable"`
+	StreamRequestFrequency int      `toml:"stream_request_frequency"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	// Internal state
+	connection     *gomavlib.Node
+	endpointConfig []gomavlib.EndpointConf
+	terminated     bool
+}
+
+//go:embed sample.conf
+var sampleConfig string
+
+func (*Mavlink) SampleConfig() string {
+	return sampleConfig
+}
+
+func (s *Mavlink) Init() error {
+	// Parse out the Mavlink endpoint.
+	endpointConfig, err := ParseMavlinkEndpointConfig(s.FcuURL)
+	if err != nil {
+		return err
+	}
+	s.endpointConfig = endpointConfig
+
+	return nil
+}
+
+func (s *Mavlink) Start(acc telegraf.Accumulator) error {
+	// Start MAVLink endpoint
+	connection, err := gomavlib.NewNode(gomavlib.NodeConf{
+		Endpoints:              s.endpointConfig,
+		Dialect:                ardupilotmega.Dialect,
+		OutVersion:             gomavlib.V2,
+		OutSystemID:            s.SystemID,
+		StreamRequestEnable:    s.StreamRequestEnable,
+		StreamRequestFrequency: s.StreamRequestFrequency,
+	})
+	if err != nil {
+		return &internal.StartupError{
+			Err:   fmt.Errorf("mavlink client failed: %w", err),
+			Retry: true,
+		}
+	}
+	s.terminated = false
+	s.connection = connection
+
+	// Start routine to connect to Mavlink and stream out data async
+	go func() {
+		defer s.connection.Close()
+		if s.terminated {
+			return
+		}
+
+		// Process MAVLink messages
+		// Use reflection to retrieve and handle all message types.
+		// (There are several hundred Mavlink message types)
+		for evt := range s.connection.Events() {
+			if s.terminated {
+				return
+			}
+			switch evt := evt.(type) {
+			case *gomavlib.EventFrame:
+				result := MavlinkEventFrameToMetric(evt)
+				if len(s.MessageFilter) > 0 && !choice.Contains(result.Name(), s.MessageFilter) {
+					continue
+				}
+				result.AddTag("fcu_url", s.FcuURL)
+				acc.AddMetric(result)
+
+			case *gomavlib.EventChannelOpen:
+				s.Log.Debugf("Mavlink channel opened")
+
+			case *gomavlib.EventChannelClose:
+				s.Log.Debugf("Mavlink channel closed")
```

Comment:
Why do we need those? Should we do something about it if the channel is opened/closed?

Reply:
These messages indicate that the Mavlink connection has started/stopped, which is separate from the Mavlink client initializing its background thread. 

There's nothing Telegraf can do in this case but it helps the user debug the Mavlink setup which can be difficult. One example of where this could occur is if you had two TCP client connections with the same `system_id` one would get kicked off. Another example is a flaky USB port.

Without these messages it's hard to tell the difference between (1) Mavlink is not connected vs (2) Robot has a problem and is not broadcasting messages

Reply:
Would it help to reconnect if the channel is closed?